src/SuperVault.sol

```diff
@@ -430,54 +441,127 @@ contract SuperVault is BaseStrategy, ISuperVault {
+        vars.assetBalances = new uint256[](_numberOfSuperforms_);
+
+        // 1. Snapshot assets and SP balances. This is relevant for withdraws and must be calculated in advance
+        for (uint256 i; i < _numberOfSuperforms_; ++i) {
+            (superform,,) = _superformIds_[i].getSuperform();
+            superformContract = IBaseForm(superform);
+
+            vars.spBalances[i] =
+                ISuperPositions(_getAddress(keccak256("SUPER_POSITIONS"))).balanceOf(address(this), _superformIds_[i]);
+            vars.assetBalances[i] = superformContract.previewRedeemFrom(vars.spBalances[i]);
+            vars.totalAssetsInVaults += vars.assetBalances[i];
+        }
+        bytes[] memory dataToEncode = new bytes[](_numberOfSuperforms_);
+
+        // 2. Add logic for deposit/withdraw cases to calculate amounts
+        for (uint256 i; i < _numberOfSuperforms_; ++i) {
             mvData.liqRequests[i].token = address(asset);
 
             (superform,,) = mvData.superformIds[i].getSuperform();
             superformContract = IBaseForm(superform);
 
-            if (isDeposit) {
+            if (isDeposit_) {
+                dataToEncode[i] = _prepareDepositExtraFormDataForSuperform(mvData.superformIds[i]);
+
                 /// @notice rounding down to avoid one-off issue
                 mvData.amounts[i] = amount_.mulDiv(weights[i], TOTAL_WEIGHT, Math.Rounding.Down);
                 mvData.outputAmounts[i] = superformContract.previewDepositTo(mvData.amounts[i]);
             } else {
-                /// @dev assets
-                mvData.outputAmounts[i] = amount_.mulDiv(weights[i], TOTAL_WEIGHT, Math.Rounding.Down);
-                /// @dev shares - in 4626Form this uses convertToShares in 5115Form this uses previewDeposit
-                mvData.amounts[i] = superformContract.previewDepositTo(mvData.outputAmounts[i]);
+                // Check if vault is ERC5115
+                bool isERC5115 = _isERC5115Vault(mvData.superformIds[i]);
+
+                if (isERC5115) {
+                    mvData.liqRequests[i].interimToken = address(asset);
+                }
+
+                if (amount_ >= vars.totalAssetsInVaults) {
+                    // If withdrawing everything, use full SP balance
+                    mvData.amounts[i] = vars.spBalances[i];
+
+                    // If vault is ERC5115, subtract tolerance constant because of minAmountOut check to avoid errors.
+                    // This is slippage protected anyway at the form level
+                    mvData.outputAmounts[i] =
+                        isERC5115 ? vars.assetBalances[i] - TOLERANCE_CONSTANT : vars.assetBalances[i];
+                } else {
+                    // For partial withdrawals, use proportional amounts
+                    uint256 amountOut = amount_.mulDiv(weights[i], TOTAL_WEIGHT, Math.Rounding.Down);
+
+                    mvData.outputAmounts[i] = isERC5115 ? amountOut - TOLERANCE_CONSTANT : amountOut;
```

Comment:
perhaps have a function to adjust for the tolerance, for easier maintenance, something like:
```solidity
function tolerance(bool isERC5115, uint256 amount) view returns (uint256){
   return isERC5115 ? amount - TOLERANCE_CONSTANT : amount;
}
```

Reply:
[30f5f82](https://github.com/superform-xyz/SuperVaults/pull/36/commits/30f5f822cc7278e2a841739d1c025252f0fdbd83) added